File: src/big_config/workflow.clj
(ns big-config.workflow
  (:require [big-config.core :as core]
            [big-config.git :as git]
            [big-config.keys :as k]
            [big-config.lock :as lock]
            [big-config.pluggable :as pluggable]
            [big-config.render :as render]
            [big-config.run :as run]
            [big-config.unlock :as unlock]
            [big-config.utils :as u]
            [clojure.string :as str]))

(def start :big-config.workflow/start)
(def end :big-config.workflow/end)
(def lock-step :big-config.workflow/lock)
(def git-check :big-config.workflow/git-check)
(def render-step :big-config.workflow/render)
(def create :big-config.workflow/create)
(def delete :big-config.workflow/delete)
(def validate :big-config.workflow/validate)
(def describe :big-config.workflow/describe)
(def exec :big-config.workflow/exec)
(def git-push :big-config.workflow/git-push)
(def unlock-any :big-config.workflow/unlock-any)

(def print-step-fn
  (core/create-step-fn
   {:before-f (fn [step opts]
                (let [failed? (and (contains? opts k/exit) (not= 0 (get opts k/exit)))
                      prefix (if failed? "✖" "➜")
                      msg (case step
                            :big-config.workflow/lock (str "Lock (owner " (or (get opts k/lock-owner) "") ")")
                            :big-config.workflow/unlock-any "Unlock any"
                            :big-config.workflow/git-check "Checking if the working directory is clean"
                            :big-config.workflow/render (str "Rendering workflow: " (or (get opts k/wf-name) ""))
                            :big-config.run/run-cmd (str "Running:\n> " (or (first (get opts k/run-cmds [])) ""))
                            nil)]
                  (when msg
                    (binding [*out* *err*]
                      (println prefix msg)))))
    :after-f (fn [step opts]
               (when (and (pos? (get opts k/exit 0)) (#{:big-config.workflow/git-check :big-config.run/run-cmd} step))
                 (binding [*out* *err*]
                   (println "✖" (if (= step :big-config.workflow/git-check)
                                  "Working directory is NOT clean"
                                  (str "Failed running:\n> " (or (first (get opts k/run-cmds [])) "")))))))}))

(defn- resolve-fn [key opts & [default-value]]
  (if-let [f (get opts key)]
    (u/->fn f)
    (if (some? default-value)
      default-value
      (throw (ex-info (str "`" key "` not defined") opts)))))

(defn select-globals [opts]
  (let [globals (get opts :globals [k/env k/run-shell-opts k/render-module k/render-profile
                                    k/wf-prefix k/wf-object-prefix :globals])]
    (select-keys opts globals)))

(defn run-steps [step-fns opts]
  (let [globals-opts (select-globals opts)
        create-opts (merge (get opts k/wf-create-opts {}) globals-opts)
        delete-opts (merge (get opts k/wf-delete-opts {}) globals-opts)
        opts* (atom opts)
        queued-steps (atom (mapv #(if (k/namespace-of %) (k/normalize-keyword %) (k/qualify k/workflow-ns %))
                                  (get opts k/wf-steps [])))
        wf (pluggable/create-workflow-star
            {:first-step start
             :last-step end
             :wire-fn (fn [step]
                        (case step
                          :big-config.workflow/start [core/ok nil]
                          :big-config.workflow/lock [(fn [o] (lock/lock step-fns o)) nil]
                          :big-config.workflow/git-check [(fn [o] (git/check step-fns o)) nil]
                          :big-config.workflow/render [(fn [o] (render/templates step-fns o)) nil]
                          :big-config.workflow/create [(fn [o] ((resolve-fn k/wf-create-fn opts) step-fns o)) nil]
                          :big-config.workflow/delete [(fn [o] ((resolve-fn k/wf-delete-fn opts) step-fns o)) nil]
                          :big-config.workflow/validate [(fn [o] ((resolve-fn k/wf-validate-fn opts (fn [_s x] (core/ok x))) step-fns o)) nil]
                          :big-config.workflow/describe [(fn [o] ((resolve-fn k/wf-describe-fn opts (fn [_s x] (core/ok x))) step-fns o)) nil]
                          :big-config.workflow/exec [(fn [o] (run/run-cmds step-fns o)) nil]
                          :big-config.workflow/git-push [git/git-push nil]
                          :big-config.workflow/unlock-any [(fn [o] (unlock/unlock-any step-fns o)) nil]
                          :big-config.workflow/end [identity nil]
                          [identity nil]))
             :next-fn (fn [step _next-step step-opts]
                        (if (#{create delete} step)
                          (swap! opts* #(assoc %
                                                k/exit (get step-opts k/exit)
                                                k/err (get step-opts k/err)
                                                step (conj (vec (get % step [])) step-opts)))
                          (reset! opts* step-opts))
                        (cond
                          (= step end) [nil @opts*]
                          (pos? (get step-opts k/exit 0)) [end @opts*]
                          :else (let [next-step (first @queued-steps)]
                                  (swap! queued-steps #(vec (rest %)))
                                  (if next-step
                                    [next-step (cond
                                                 (= next-step create) create-opts
                                                 (= next-step delete) delete-opts
                                                 :else @opts*)]
                                    [end @opts*]))))})]
    (wf step-fns @opts*)))

(def parse-arg-steps #{"lock" "git-check" "render" "create" "delete" "validate" "describe" "exec" "git-push" "unlock-any"})
(def ^:dynamic *parse-args-steps* parse-arg-steps)

(defn- tokenize [str-or-args]
  (cond
    (string? str-or-args) (let [trimmed (str/trim str-or-args)]
                            (if (str/blank? trimmed) [] (str/split trimmed #"\s+")))
    (nil? str-or-args) []
    :else (mapv str str-or-args)))

(defn parse-args [str-or-args]
  (loop [xs (vec (tokenize str-or-args))
         steps []
         cmds []]
    (if (empty? xs)
      {k/wf-steps steps k/run-cmds cmds}
      (let [token (first xs)
            token-kw (k/normalize-keyword token)
            token-name (k/name-of token-kw)]
        (cond
          (and (contains? parse-arg-steps token-name) (nil? (k/namespace-of token-kw)))
          (recur (subvec xs 1) (conj steps (keyword token-name)) cmds)

          (= token "--")
          (let [rest-args (subvec xs 1)]
            (when (empty? rest-args)
              (throw (ex-info "-- cannot be without a command" {})))
            {k/wf-steps (cond-> steps (not (some #{:exec} steps)) (conj :exec))
             k/run-cmds (conj cmds (str/join " " rest-args))})

          :else
          (recur (subvec xs 1)
                 (cond-> steps (not (some #{:exec} steps)) (conj :exec))
                 (conj cmds (str/replace token ":" " "))))))))

(defn- parse-path [path]
  (vec (remove str/blank? (str/split (str path) #"/"))))

(defn- build-path [parts profile suffix]
  (str/join "/" (concat parts [(str profile "-" suffix)])))

(defn new-prefix [opts first-step]
  (let [prefix (or (get opts k/wf-prefix) ".dist")
        object-prefix (or (get opts k/wf-object-prefix) "tofu")
        profile (or (get opts k/render-profile) "default")
        dirs (parse-path prefix)
        object-dirs (parse-path object-prefix)
        last-dir (or (peek dirs) "")
        profile-found? (str/starts-with? last-dir (str profile))
        prev-hash (if profile-found? (last (str/split last-dir #"-")) "")
        base-dirs (if profile-found? (pop dirs) dirs)
        base-object-dirs (if profile-found? (pop object-dirs) object-dirs)
        suffix (u/hash-string (str (k/key-string first-step) prev-hash) 8)]
    (assoc opts
           k/wf-prefix (build-path base-dirs profile suffix)
           k/wf-object-prefix (build-path base-object-dirs profile suffix))))

(defn path [opts name]
  (str (or (get opts k/wf-prefix) ".dist") "/" (u/keyword->path name)))

(defn prepare [opts overrides]
  (u/assert-args-present {:opts opts :overrides overrides :name (get opts k/wf-name)})
  (let [prefix (get overrides k/wf-prefix)
        object-prefix (get overrides k/wf-object-prefix)
        path-fn (or (get overrides k/wf-path-fn)
                    (fn [o] (str (or prefix ".dist") "/" (u/keyword->path (get o k/wf-name)))))
        object-fn (or (get overrides k/wf-object-fn)
                      (fn [o] (str (or object-prefix "tofu") "/" (u/keyword->name (get o k/wf-name)))))
        merged (merge opts overrides)
        dir (path-fn merged)
        object (object-fn merged)
        params (get overrides k/wf-params {})
        templates (mapv #(merge % params {:target-dir dir :target-object object})
                        (get merged k/render-templates []))]
    (assoc merged
           k/render-templates templates
           k/run-shell-opts (merge (get merged k/run-shell-opts {}) {:dir dir}))))

(defn merge-params [tools params opts]
  (reduce (fn [out tool]
            (reduce (fn [out root]
                      (let [existing (get-in out [root tool k/wf-params] {})]
                        (assoc-in out [root tool k/wf-params] (merge params existing))))
                    out
                    [k/wf-create-opts k/wf-delete-opts]))
          opts
          tools))

(def env-prefix "BC_PAR_")

(defn read-bc-pars
  ([opts] (read-bc-pars opts (System/getenv)))
  ([opts env]
   (let [params-from-env (into {}
                               (for [[ek ev] env
                                     :let [ek (name ek)]
                                     :when (and (str/starts-with? ek env-prefix) (some? ev))]
                                 [(-> ek
                                      (subs (count env-prefix))
                                      str/lower-case
                                      (str/replace "_" "-")
                                      (str/replace "." "-")
                                      keyword)
                                  ev]))]
     (assoc opts k/wf-params (merge (get opts k/wf-params {}) params-from-env)))))

(defonce workflow-registry (atom {}))

(defn register-workflow [name workflow]
  (swap! workflow-registry assoc (k/normalize-keyword name) workflow))

(defn unregister-workflow [name]
  (swap! workflow-registry dissoc (k/normalize-keyword name)))

(defn create-workflow-star [{:keys [first-step last-step pipeline firstStep lastStep]}]
  (let [first-step (k/normalize-keyword (or first-step firstStep))
        last-step (some-> (or last-step lastStep) k/normalize-keyword)]
    (when-not first-step
      (throw (IllegalArgumentException. ":first-step is required")))
    (when-not (vector? pipeline)
      (throw (IllegalArgumentException. ":pipeline must be like [::tool/tofu ...\n::tool/ansible ...")))
    (fn workflow*
      ([] [first-step (or last-step (if-let [ns (k/namespace-of first-step)] (keyword ns "end") :end))])
      ([step-fns opts]
       (let [actual-last-step (or last-step (if-let [ns (k/namespace-of first-step)] (keyword ns "end") :end))
             globals-opts (new-prefix (select-globals opts) first-step)
             pairs (partition 2 pipeline)
             step->opts-and-fn (into {}
                                      (for [[step tuple] pairs
                                            :let [step (k/normalize-keyword step)
                                                  tuple (if (sequential? tuple) (vec tuple) [tuple])
                                                  args (or (first tuple) [])
                                                  opts-fn (if (nil? (second tuple)) identity (u/->fn (second tuple)))
                                                  step-opts (get opts (k/add-suffix step "-opts") {})]]
                                        [step [(merge (parse-args args) globals-opts step-opts) opts-fn]]))
             steps (mapv (comp k/normalize-keyword first) pairs)
             steps-set (set steps)
             sequence (vec (concat [first-step] steps [actual-last-step nil]))
             step->next (into {} (map vector sequence (rest sequence)))
             opts* (atom opts)
             step->fn (fn [step]
                        (cond
                          (= step first-step) core/ok
                          (= step actual-last-step) identity
                          :else (let [registered (or (get @workflow-registry step) (get opts step))]
                                  (when-not (fn? registered)
                                    (throw (ex-info (str "Workflow '" step "' is not registered") {:step step})))
                                  (fn [o] (registered step-fns o)))))
             wf (pluggable/create-workflow-star
                 {:first-step first-step
                  :last-step actual-last-step
                  :wire-fn (fn [step] [(step->fn step) (get step->next step)])
                  :next-fn (fn [step next-step step-opts]
                             (if (contains? steps-set step)
                               (reset! opts* (assoc @opts*
                                                    k/exit (get step-opts k/exit)
                                                    k/err (get step-opts k/err)
                                                    step step-opts))
                               (reset! opts* step-opts))
                             (cond
                               (or (= step actual-last-step) (= step end)) [nil @opts*]
                               (pos? (get step-opts k/exit 0)) [actual-last-step @opts*]
                               :else (let [[new-opts opts-fn] (get step->opts-and-fn next-step [@opts* identity])]
                                       [next-step (opts-fn new-opts)])))})]
         (wf step-fns opts))))))

(def ->workflow* create-workflow-star)

(u/register-function 'big-config.workflow/run-steps run-steps)
(u/register-function 'big-config.workflow/parse-args parse-args)
(u/register-function 'big-config.workflow/prepare prepare)

File: src/big_config/workflow.py
from __future__ import annotations

import hashlib
import os
import sys
from collections.abc import Callable, Iterable, Mapping, Sequence
from typing import Any

from . import ENV, ERR, EXIT
from . import core
from . import git, lock, pluggable, render, run, unlock
from .utils import BigConfigError, has_namespace, keyword, keyword_to_name, keyword_to_path, name, namespace, requiring_resolve, to_fn

STEPS = "big-config.workflow/steps"
CREATE_FN = "big-config.workflow/create-fn"
BUILD_FN = "big-config.workflow/build-fn"
DELETE_FN = "big-config.workflow/delete-fn"
VALIDATE_FN = "big-config.workflow/validate-fn"
DESCRIBE_FN = "big-config.workflow/describe-fn"
CREATE_OPTS = "big-config.workflow/create-opts"
BUILD_OPTS = "big-config.workflow/build-opts"
DELETE_OPTS = "big-config.workflow/delete-opts"
NAME = "big-config.workflow/name"
PATH_FN = "big-config.workflow/path-fn"
OBJECT_FN = "big-config.workflow/object-fn"
PREFIX = "big-config.workflow/prefix"
OBJECT_PREFIX = "big-config.workflow/object-prefix"
PARAMS = "big-config.workflow/params"
GLOBALS = "big-config.workflow/globals"

PARSE_ARGS_STEPS: set[str] = {"lock", "git-check", "render", "create", "build", "delete", "validate", "describe", "exec", "git-push", "unlock-any"}
_WORKFLOW_REGISTRY: dict[str, Callable[[list[Any], dict[str, Any]], dict[str, Any]]] = {}


def _print_step_before(step: str, opts: Mapping[str, Any]) -> None:
    failed = opts.get(EXIT) is not None and opts.get(EXIT) != 0
    prefix = "✖" if failed else "➜"
    msg = None
    if step == "big-config.workflow/lock":
        msg = f"Lock (owner {opts.get('big-config.lock/owner', '')})"
    elif step == "big-config.workflow/unlock-any":
        msg = "Unlock any"
    elif step == "big-config.workflow/git-check":
        msg = "Checking if the working directory is clean"
    elif step == "big-config.workflow/render":
        msg = f"Rendering workflow: {opts.get(NAME, '')}"
    elif step == "big-config.run/run-cmd":
        msg = f"Running:\n> {(opts.get(run.CMDS) or [''])[0] or ''}"
    if msg:
        print(f"{prefix} {msg}", file=sys.stderr)


def _print_step_after(step: str, opts: Mapping[str, Any]) -> None:
    if opts.get(EXIT, 0) > 0 and step in {"big-config.workflow/git-check", "big-config.run/run-cmd"}:
        msg = "Working directory is NOT clean" if step == "big-config.workflow/git-check" else f"Failed running:\n> {(opts.get(run.CMDS) or [''])[0] or ''}"
        print(f"✖ {msg}", file=sys.stderr)


print_step_fn = core.step_fn({"before_f": _print_step_before, "after_f": _print_step_after})


def register_workflow_step(step: str, fn: Callable[[list[Any], dict[str, Any]], dict[str, Any]]) -> None:
    _WORKFLOW_REGISTRY[step] = fn


def unregister_workflow_step(step: str) -> None:
    _WORKFLOW_REGISTRY.pop(step, None)


def _resolve_fn(key: str, opts: Mapping[str, Any], default: Any = None, required: bool = True) -> Callable[..., Any]:
    value = opts.get(key)
    if value is None:
        if required and default is None:
            raise BigConfigError(f"`{key}` not defined", opts)
        return default
    return to_fn(value)


def select_globals(opts: Mapping[str, Any]) -> dict[str, Any]:
    selected = opts.get("globals")
    keys = selected or [ENV, run.SHELL_OPTS, render.MODULE, render.PROFILE, PREFIX, OBJECT_PREFIX, GLOBALS]
    return {k: opts[k] for k in keys if k in opts}


def _qualify_workflow_step(step: str) -> str:
    step = str(step).lstrip(":")
    return step if has_namespace(step) else keyword("big-config.workflow", step)


def _hook_or_ok(key: str, opts: Mapping[str, Any]) -> Callable[[list[Any], dict[str, Any]], dict[str, Any]]:
    value = opts.get(key)
    if value is None:
        return lambda _step_fns, hook_opts: core.ok(hook_opts)
    return to_fn(value)


def run_steps(step_fns: list[Any], opts: Mapping[str, Any]) -> dict[str, Any]:
    globals_opts = select_globals(opts)
    create_opts = {**dict(opts.get(CREATE_OPTS, {}) or {}), **globals_opts}
    build_opts = {**dict(opts.get(BUILD_OPTS, {}) or {}), **globals_opts}
    delete_opts = {**dict(opts.get(DELETE_OPTS, {}) or {}), **globals_opts}
    opts_acc: dict[str, Any] = dict(opts)
    steps_queue = [_qualify_workflow_step(step) for step in opts.get(STEPS, [])]

    def wire_fn(step: str, resolved_step_fns: list[Callable[..., Any]]):
        if step == "big-config.workflow/start":
            return core.ok, None
        if step == "big-config.workflow/lock":
            return lambda inner: lock.lock(resolved_step_fns, inner), None
        if step == "big-config.workflow/git-check":
            return lambda inner: git.check(resolved_step_fns, inner), None
        if step == "big-config.workflow/render":
            return lambda inner: render.templates(resolved_step_fns, inner), None
        if step == "big-config.workflow/create":
            return lambda inner: _resolve_fn(CREATE_FN, opts)(resolved_step_fns, inner), None
        if step == "big-config.workflow/build":
            return lambda inner: _resolve_fn(BUILD_FN, opts)(resolved_step_fns, inner), None
        if step == "big-config.workflow/delete":
            return lambda inner: _resolve_fn(DELETE_FN, opts)(resolved_step_fns, inner), None
        if step == "big-config.workflow/validate":
            return lambda inner: _hook_or_ok(VALIDATE_FN, opts)(resolved_step_fns, inner), None
        if step == "big-config.workflow/describe":
            return lambda inner: _hook_or_ok(DESCRIBE_FN, opts)(resolved_step_fns, inner), None
        if step == "big-config.workflow/exec":
            return lambda inner: run.run_cmds(resolved_step_fns, inner), None
        if step == "big-config.workflow/git-push":
            return git.git_push, None
        if step == "big-config.workflow/unlock-any":
            return lambda inner: unlock.unlock_any(resolved_step_fns, inner), None
        return lambda x: x, None

    def next_fn(step: str, _next_step: str | None, current_opts: dict[str, Any]):
        nonlocal opts_acc, steps_queue
        exit_code = current_opts.get(EXIT)
        if step in {"big-config.workflow/create", "big-config.workflow/build", "big-config.workflow/delete"}:
            opts_acc.update({k: current_opts.get(k) for k in [EXIT, ERR] if k in current_opts})
            opts_acc.setdefault(step, []).append(current_opts)
        else:
            opts_acc = current_opts
        if step == "big-config.workflow/end":
            return None, opts_acc
        if isinstance(exit_code, int) and exit_code > 0:
            return "big-config.workflow/end", opts_acc
        if steps_queue:
            next_step = steps_queue.pop(0)
            if next_step == "big-config.workflow/create":
                return next_step, dict(create_opts)
            if next_step == "big-config.workflow/build":
                return next_step, dict(build_opts)
            if next_step == "big-config.workflow/delete":
                return next_step, dict(delete_opts)
            return next_step, opts_acc
        return "big-config.workflow/end", opts_acc

    wf = pluggable.workflow_star(
        {
            "first_step": "big-config.workflow/start",
            "last_step": "big-config.workflow/end",
            "wire_fn": wire_fn,
            "next_fn": next_fn,
        }
    )
    return wf(step_fns, opts_acc)


def parse_args(str_or_args: str | Sequence[str]) -> dict[str, Any]:
    xs: list[str]
    if isinstance(str_or_args, str):
        stripped = str_or_args.strip()
        xs = stripped.split() if stripped else []
    else:
        xs = [str(x) for x in str_or_args]
    token = xs.pop(0) if xs else None
    steps: list[str] = []
    cmds: list[str] = []
    while token is not None:
        if token in {"--"}:
            if not xs:
                raise BigConfigError("-- cannot be without a command", {})
            if "exec" not in steps:
                steps.append("exec")
            cmds.append(" ".join(xs))
            xs = []
            token = None
        elif token in PARSE_ARGS_STEPS:
            steps.append(token)
            token = xs.pop(0) if xs else None
        else:
            if "exec" not in steps:
                steps.append("exec")
            cmds.append(token.replace(":", " "))
            token = xs.pop(0) if xs else None
    return {STEPS: steps, run.CMDS: cmds}


def _add_suffix(kw: str, suffix: str) -> str:
    return keyword(namespace(kw), name(kw) + suffix)


def _parse_path(path: str) -> list[str]:
    return str(path).split("/")


def _build_path(parts: Iterable[str], profile: str, suffix: str) -> str:
    return "/".join([*list(parts), f"{profile}-{suffix}"])


def new_prefix(opts: Mapping[str, Any], first_step: str) -> dict[str, Any]:
    prefix = str(opts.get(PREFIX, ".dist"))
    object_prefix = str(opts.get(OBJECT_PREFIX, "tofu"))
    profile = str(opts.get(render.PROFILE, "default"))
    dirs = _parse_path(prefix)
    object_dirs = _parse_path(object_prefix)
    profile_found = bool(dirs and dirs[-1].startswith(profile))
    prev_hash = dirs[-1].split("-")[-1] if profile_found else ""
    base_dirs = dirs[:-1] if profile_found else dirs
    base_object_dirs = object_dirs[:-1] if profile_found else object_dirs
    digest = hashlib.sha256(f"{first_step}{prev_hash}".encode()).hexdigest()[:8]
    new_opts = dict(opts)
    new_opts[PREFIX] = _build_path(base_dirs, profile, digest)
    new_opts[OBJECT_PREFIX] = _build_path(base_object_dirs, profile, digest)
    return new_opts


def _pipeline_pairs(pipeline: Sequence[Any]) -> list[tuple[str, Sequence[Any]]]:
    if len(pipeline) % 2 != 0:
        raise ValueError(":pipeline must contain alternating step and [args opts_fn] entries")
    return [(str(pipeline[i]).lstrip(":"), pipeline[i + 1]) for i in range(0, len(pipeline), 2)]


def _resolve_pipeline_fn(step: str, step_fns: list[Any]) -> Callable[[dict[str, Any]], dict[str, Any]]:
    if step in _WORKFLOW_REGISTRY:
        return lambda opts: _WORKFLOW_REGISTRY[step](step_fns, opts)
    ns = namespace(step)
    nm = name(step).replace("-", "_")
    if ns is None:
        raise BigConfigError("Pipeline step must be qualified", {"step": step})
    fn = requiring_resolve(f"{ns}/{nm}")
    return lambda opts: fn(step_fns, opts)


def workflow_star(wf_opts: Mapping[str, Any]) -> Callable[[list[Any], Mapping[str, Any]], dict[str, Any]]:
    pipeline = wf_opts.get("pipeline")
    if not isinstance(pipeline, Sequence) or isinstance(pipeline, (str, bytes, bytearray, Mapping)):
        raise ValueError(":pipeline must be like [::tool/tofu ...\n::tool/ansible ...")
    first_step = str(wf_opts.get("first_step", wf_opts.get("first-step"))).lstrip(":")
    last_step0 = wf_opts.get("last_step", wf_opts.get("last-step"))
    last_step = str(last_step0).lstrip(":") if last_step0 else keyword(namespace(first_step), "end")
    pairs = _pipeline_pairs(pipeline)

    def run_wf(step_fns: list[Any], opts0: Mapping[str, Any]) -> dict[str, Any]:
        globals_opts = new_prefix(select_globals(opts0), first_step)
        step_to_opts_and_opts_fn: dict[str, tuple[dict[str, Any], Callable[[dict[str, Any]], dict[str, Any]]]] = {}
        for step, args_and_maybe_fn in pairs:
            args_list = list(args_and_maybe_fn)
            args = args_list[0] if args_list else []
            opts_fn = to_fn(args_list[1], lambda x: x) if len(args_list) > 1 else (lambda x: x)
            step_opts = dict(opts0.get(_add_suffix(step, "-opts"), {}) or {})
            step_args = parse_args(args)
            step_to_opts_and_opts_fn[step] = ({**step_args, **globals_opts, **step_opts}, opts_fn)

        sequence = [first_step, *[step for step, _ in pairs], last_step]
        step_to_f_and_next: dict[str, tuple[Callable[[dict[str, Any]], dict[str, Any]], str | None]] = {}
        for current, nxt in zip(sequence, [*sequence[1:], None]):
            if current == first_step:
                f = core.ok
            elif current == last_step:
                f = lambda x: x
            else:
                f = _resolve_pipeline_fn(current, step_fns)
            step_to_f_and_next[current] = (f, nxt)

        opts_acc: dict[str, Any] = dict(opts0)
        steps_set = {step for step, _ in pairs}

        def wire_fn(step: str, _resolved_step_fns: list[Callable[..., Any]]):
            return step_to_f_and_next[step]

        def next_fn(step: str, next_step: str | None, current_opts: dict[str, Any]):
            nonlocal opts_acc
            exit_code = current_opts.get(EXIT)
            if step in steps_set:
                opts_acc.update({k: current_opts.get(k) for k in [EXIT, ERR] if k in current_opts})
                opts_acc[step] = current_opts
            else:
                opts_acc = current_opts
            if step == last_step or next_step is None:
                return None, opts_acc
            if isinstance(exit_code, int) and exit_code > 0:
                return last_step, opts_acc
            new_opts, opts_fn = step_to_opts_and_opts_fn.get(next_step, (opts_acc, lambda x: x))
            return next_step, opts_fn(dict(new_opts))

        wf = pluggable.workflow_star({"first_step": first_step, "last_step": last_step, "wire_fn": wire_fn, "next_fn": next_fn})
        return wf(step_fns, dict(opts0))

    return run_wf


def path(opts: Mapping[str, Any], wf_name: str) -> str:
    return f"{opts.get(PREFIX, '.dist')}/{keyword_to_path(wf_name)}"


def prepare(opts: Mapping[str, Any], overrides: Mapping[str, Any]) -> dict[str, Any]:
    if opts.get(NAME) is None:
        raise ValueError("Argument name is nil")
    prefix = overrides.get(PREFIX)
    object_prefix = overrides.get(OBJECT_PREFIX)
    path_fn = overrides.get(PATH_FN) or (lambda current: f"{current.get(PREFIX, prefix or '.dist')}/{keyword_to_path(current[NAME])}")
    object_fn = overrides.get(OBJECT_FN) or (lambda current: f"{current.get(OBJECT_PREFIX, object_prefix or 'tofu')}/{keyword_to_name(current[NAME])}")
    merged = {**dict(opts), **dict(overrides)}
    target_dir = path_fn(merged)
    target_object = object_fn(merged)
    params = dict(overrides.get(PARAMS, {}) or {})
    templates = []
    for template in merged.get(render.TEMPLATES, []) or []:
        templates.append({**dict(template), **params, "target-dir": target_dir, "target-object": target_object})
    shell_opts = {**merged.get(run.SHELL_OPTS, {}), "dir": target_dir}
    return {**merged, render.TEMPLATES: templates, run.SHELL_OPTS: shell_opts}


def merge_params(tools: Sequence[str], params: Mapping[str, Any], opts: Mapping[str, Any]) -> dict[str, Any]:
    result = dict(opts)
    for tool in tools:
        for root in [CREATE_OPTS, BUILD_OPTS, DELETE_OPTS]:
            root_map = dict(result.get(root, {}) or {})
            tool_map = dict(root_map.get(tool, {}) or {})
            existing = dict(tool_map.get(PARAMS, {}) or {})
            tool_map[PARAMS] = {**dict(params), **existing}
            root_map[tool] = tool_map
            result[root] = root_map
    return result


def read_bc_pars(opts: Mapping[str, Any], env: Mapping[str, str] | None = None) -> dict[str, Any]:
    env = env or os.environ
    prefix = "BC_PAR_"
    params_from_env = {
        key[len(prefix) :].lower().replace("_", "-").replace(".", "-"): value
        for key, value in env.items()
        if key.startswith(prefix)
    }
    merged_params = {**dict(opts.get(PARAMS, {}) or {}), **params_from_env}
    return {**dict(opts), PARAMS: merged_params}


# Compatibility alias.
globals()["->workflow*"] = workflow_star
